### DISCUSSION TITLE
Installing a custom kiali version from a private image registry

### DISCUSSION THREAD

**Discussion by vaidx** [kiali/kiali#3864] [Q&A]
I wish to install Kiali CR v1.31 using the Kiali Operator v1.32. The image is stored in a different registry.
The `spec.deployment.image_name = "private registry link"` and spec.deployment.image_version = "v1.31". 
But Kiali doesn't get installed in the istio-system namespace and instead throws the below errors during reconciliation on the step of creating a Deployment for kiali.

**Describe the bug**
`2021-04-09T02:06:52.688232506Z fatal: [localhost]: FAILED! => {"attempts": 6, "changed": false, "module_stderr": "Traceback (most recent call last):\n File \"/opt/ansible/.ansible/tmp/ansible-tmp-1617934011.7817597-58944-82795733936648/AnsiballZ_k8s.py\", line 102, in <module>\n _ansiballz_main()\n File \"/opt/ansible/.ansible/tmp/ansible-tmp-1617934011.7817597-58944-82795733936648/AnsiballZ_k8s.py\", line 94, in _ansiballz_main\n invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n File \"/opt/ansible/.ansible/tmp/ansible-tmp-1617934011.7817597-58944-82795733936648/AnsiballZ_k8s.py\", line 40, in invoke_module\n runpy.run_module(mod_name='ansible.modules.clustering.k8s.k8s', init_globals=None, run_name='__main__', alter_sys=True)\n File \"/usr/lib64/python3.6/runpy.py\", line 205, in run_module\n return _run_module_code(code, init_globals, run_name, mod_spec)\n File \"/usr/lib64/python3.6/runpy.py\", line 96, in _run_module_code\n mod_name, mod_spec, pkg_name, script_name)\n File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n exec(code, run_globals)\n File \"/tmp/ansible_k8s_payload_a8nlajen/ansible_k8s_payload.zip/ansible/modules/clustering/k8s/k8s.py\", line 281, in <module>\n File \"/tmp/ansible_k8s_payload_a8nlajen/ansible_k8s_payload.zip/ansible/modules/clustering/k8s/k8s.py\", line 277, in main\n File \"/tmp/ansible_k8s_payload_a8nlajen/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py\", line 145, in __init__\n File \"/tmp/ansible_k8s_payload_a8nlajen/ansible_k8s_payload.zip/ansible/module_utils/k8s/raw.py\", line 145, in <listcomp>\n File \"/usr/lib64/python3.6/site-packages/yaml/__init__.py\", line 84, in load_all\n yield loader.get_data()\n File \"/usr/lib64/python3.6/site-packages/yaml/constructor.py\", line 31, in get_data\n return self.construct_document(self.get_node())\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 27, in get_node\n return self.compose_document()\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 55, in compose_document\n node = self.compose_node(None, None)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 84, in compose_node\n node = self.compose_mapping_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 133, in compose_mapping_node\n item_value = self.compose_node(node, item_key)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 84, in compose_node\n node = self.compose_mapping_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 133, in compose_mapping_node\n item_value = self.compose_node(node, item_key)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 84, in compose_node\n node = self.compose_mapping_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 133, in compose_mapping_node\n item_value = self.compose_node(node, item_key)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 84, in compose_node\n node = self.compose_mapping_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 133, in compose_mapping_node\n item_value = self.compose_node(node, item_key)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 82, in compose_node\n node = self.compose_sequence_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 111, in compose_sequence_node\n node.value.append(self.compose_node(node, index))\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 84, in compose_node\n node = self.compose_mapping_node(anchor)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 133, in compose_mapping_node\n item_value = self.compose_node(node, item_key)\n File \"/usr/lib64/python3.6/site-packages/yaml/composer.py\", line 64, in compose_node\n if self.check_event(AliasEvent):\n File \"/usr/lib64/python3.6/site-packages/yaml/parser.py\", line 98, in check_event\n self.current_event = self.state()\n File \"/usr/lib64/python3.6/site-packages/yaml/parser.py\", line 449, in parse_block_mapping_value\n if not self.check_token(KeyToken, ValueToken, BlockEndToken):\n File \"/usr/lib64/python3.6/site-packages/yaml/scanner.py\", line 116, in check_token\n self.fetch_more_tokens()\n File \"/usr/lib64/python3.6/site-packages/yaml/scanner.py\", line 212, in fetch_more_tokens\n return self.fetch_block_entry()\n File \"/usr/lib64/python3.6/site-packages/yaml/scanner.py\", line 490, in fetch_block_entry\n self.get_mark())\nyaml.scanner.ScannerError: sequence entries are not allowed here\n in \"<unicode string>\", line 49, column 15:\n - name: -\n ^\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}
`

**Versions used**

Kiali: v1.31
Kiali Operator: v1.32
Istio: 1.81
Kubernetes flavour and version: 1.18.8-aliyun.1
--

**To Reproduce**
1. Install the kiali operator and with `cr.create` set to true.
2. Specify the `cr.spec.deployment.image_name` and `cr.spec.deployment.image_version` in values.yaml accordingly.
3. Check logs for error.

**Expected behavior**
Kiali Operator should install the Kiali CR v1.31 in the istio-system namespace.

**Comment by jmazzitelli** [comment]
Please provide the full values.yaml you are using and the full helm command you are using to install

**Comment by jmazzitelli** [comment]
This works for me:

```
kubectl create ns kiali-operator ; helm install --namespace kiali-operator --set cr.create=true --set cr.spec.deployment.image_name=fooimagename --set cr.spec.deployment.image_version=barimagever --repo https://kiali.org/helm-charts kiali-operator kiali-operator
```

results in a Kiali CR of:

```
$ kubectl get kiali kiali -n kiali-operator -o jsonpath='{.spec.deployment}'
{"accessible_namespaces":["**"],"image_name":"fooimagename","image_version":"barimagever"}
```
and the Kiali Deployment has the correct container name:

```
kubectl get deployment kiali -n kiali-operator -o jsonpath='{.spec.template.spec.containers[0].image}'
fooimagename:barimagever
```

With no errors in the operator logs.

This does not look like a bug. Without full replication procedures (what your values are and what the full command was used to install) I will assume this is user error and not classify this as a bug.

**Comment by vaidx** [comment]
Okay, it works finally now. Probably an issue with YAML and got removed after specifying the fields in the helm commands. Thanks a lot!

**Comment by jmazzitelli** [comment]
Converted this to a discussion as it is not a bug but rather helps answer the question "How to install a custom kiali version from a private image registry?"
